File: LeetCode/0729. My Calendar I/solution.py
"""
107 / 107 test cases passed.
Runtime: 236 ms
Memory Usage: 16.1 MB
"""
from sortedcontainers import SortedDict

class MyCalendar:

    def __init__(self):
        self.booked = SortedDict()

    def book(self, start: int, end: int) -> bool:
        i = self.booked.bisect_left(end)
        if i == 0 or self.booked.items()[i - 1][1] <= start:
            self.booked[start] = end
            return True
        return False

# Your MyCalendar object will be instantiated and called as such:
# obj = MyCalendar()
# param_1 = obj.book(start,end)

File: LeetCode/0729. My Calendar I/solution.cpp
/**
 * 107 / 107 test cases passed.
 * Runtime: 72 ms
 * Memory Usage: 37.7 MB 
 */
class MyCalendar {
public:
    MyCalendar() = default;
    
    bool book(int start, int end) {
        auto it = _booked.lower_bound({end, 0});
        if (it == _booked.begin() || (--it)->second <= start) {
            _booked.emplace(start, end);
            return true;
        }
        return false;
    }
    
private:
    set<pair<int, int>> _booked;
};

/**
 * Your MyCalendar object will be instantiated and called as such:
 * MyCalendar* obj = new MyCalendar();
 * bool param_1 = obj->book(start,end);
 */
